Bootloader/bootloader_core.c

```diff
@@ -0,0 +1,514 @@
+            // if STOP_CMD, restart the node
+            if (bootloader_cmd == BOOTLOADER_STOP)
+            {
+                LuosBootloader_SetState(BOOTLOADER_STOP_STATE);
+            }
+            // if READY_CMD, continue BOOTLOADER process
+            if (bootloader_cmd == BOOTLOADER_READY)
+            {
+                // save binary length
+                memcpy(&nb_bytes, &bootloader_data, sizeof(uint32_t));
+                // check free space in flash
+                if (LuosBootloader_IsEnoughSpace(nb_bytes))
+                {
+                    // send READY response
+                    LuosBootloader_SendResponse(BOOTLOADER_READY_RESP);
+                    // go to HEADER state
+                    LuosBootloader_SetState(BOOTLOADER_ERASE_STATE);
+                }
+                else
+                {
+                    // send READY response
+                    LuosBootloader_SendResponse(BOOTLOADER_ERROR_SIZE);
+                }
+            }
+            break;
+
+        case BOOTLOADER_ERASE_STATE:
+
+            if (bootloader_cmd == BOOTLOADER_ERASE)
+            {
+                // erase flash memory
+                LuosBootloader_EraseMemory();
+                // send READY response
+                LuosBootloader_SendResponse(BOOTLOADER_ERASE_RESP);
+                // go to HEADER state
+                LuosBootloader_SetState(BOOTLOADER_BIN_CHUNK_STATE);
+            }
+
+            break;
+
+        case BOOTLOADER_BIN_CHUNK_STATE:
+
+            if (bootloader_cmd == BOOTLOADER_BIN_CHUNK)
+            {
+                // reset bootloader cmd to avoid looping in this portion of the code
+                bootloader_cmd = BOOTLOADER_IDLE;
+                // handle binary data
+                LuosBootloader_ProcessData();
+                // send ack to the Host
+                LuosBootloader_SendResponse(BOOTLOADER_BIN_CHUNK_RESP);
+            }
+
+            if (bootloader_cmd == BOOTLOADER_BIN_END)
+            {
+                // save the current page in flash memory
+                LuosBootloader_SaveLastData();
+
+                // send ack to the Host
+                LuosBootloader_SendResponse(BOOTLOADER_BIN_END_RESP);
+
+                // go to BIN_END state
+                LuosBootloader_SetState(BOOTLOADER_CRC_TEST_STATE);
+            }
+            break;
+
+        case BOOTLOADER_CRC_TEST_STATE:
+
+            if (bootloader_cmd == BOOTLOADER_CRC_TEST)
+            {
+                uint8_t crc = compute_crc();
+                // send ack to the Host
+                LuosBootloader_SendCrc(BOOTLOADER_CRC_RESP, crc, sizeof(uint16_t));
+
+                // go to READY state
+                LuosBootloader_SetState(BOOTLOADER_READY_STATE);
+            }
+            break;
+
+        case BOOTLOADER_STOP_STATE:
+            // save boot_mode in flash
+            LuosHAL_SetMode(APPLICATION_MODE);
+            // wait for the command to be send to all nodes
```

Comment:
Is it Gate that sends commands?

Reply:
Yes the gate sends all reboot commands. In fact this is the PC which sends this command, the gate is just a bridge between the json messages sends from the PC and the luos network, there is no intelligence in this function.